samples/s05-computeVertex/hostCode.cpp

```diff
@@ -0,0 +1,304 @@
+  // now that everything is ready: launch it ....
+  // ##################################################################
+
+  LOG("launching ...");
+
+  bool firstFrame = true;
+  double xpos = 0.f, ypos = 0.f;
+  double lastxpos, lastypos;
+  do {
+    float speed = .001f;
+    lastxpos = xpos;
+    lastypos = ypos;
+    gprtGetCursorPos(context, &xpos, &ypos);
+    if (firstFrame) {
+      lastxpos = xpos;
+      lastypos = ypos;
+    }
+    int state = gprtGetMouseButton(context, GPRT_MOUSE_BUTTON_LEFT);
+
+    // If we click the mouse, we should rotate the camera
+    if (state == GPRT_PRESS || firstFrame) {
+      firstFrame = false;
+      float4 position = {lookFrom.x, lookFrom.y, lookFrom.z, 1.f};
+      float4 pivot = {lookAt.x, lookAt.y, lookAt.z, 1.0};
+#ifndef M_PI
+#define M_PI 3.1415926f
+#endif
+
+      // step 1 : Calculate the amount of rotation given the mouse movement.
+      float deltaAngleX = (2 * M_PI / fbSize.x);
+      float deltaAngleY = (M_PI / fbSize.y);
+      float xAngle = (lastxpos - xpos) * deltaAngleX;
+      float yAngle = (lastypos - ypos) * deltaAngleY;
+
+      // step 2: Rotate the camera around the pivot point on the first axis.
+      float4x4 rotationMatrixX = rotation_matrix(rotation_quat(lookUp, xAngle));
+      position = (mul(rotationMatrixX, (position - pivot))) + pivot;
+
+      // step 3: Rotate the camera around the pivot point on the second axis.
+      float3 lookRight = cross(lookUp, normalize(pivot - position).xyz());
+      float4x4 rotationMatrixY =
+          rotation_matrix(rotation_quat(lookRight, yAngle));
+      lookFrom = ((mul(rotationMatrixY, (position - pivot))) + pivot).xyz();
+
+      // ----------- compute variable values  ------------------
+      float3 camera_pos = lookFrom;
+      float3 camera_d00 = normalize(lookAt - lookFrom);
+      float aspect = float(fbSize.x) / float(fbSize.y);
+      float3 camera_ddu =
+          cosFovy * aspect * normalize(cross(camera_d00, lookUp));
+      float3 camera_ddv = cosFovy * normalize(cross(camera_ddu, camera_d00));
+      camera_d00 -= 0.5f * camera_ddu;
+      camera_d00 -= 0.5f * camera_ddv;
+
+      // ----------- set variables  ----------------------------
+      gprtRayGenSet3fv(rayGen, "camera.pos", (float *)&camera_pos);
+      gprtRayGenSet3fv(rayGen, "camera.dir_00", (float *)&camera_d00);
+      gprtRayGenSet3fv(rayGen, "camera.dir_du", (float *)&camera_ddu);
+      gprtRayGenSet3fv(rayGen, "camera.dir_dv", (float *)&camera_ddv);
+
+      // Use this to upload all set parameters to our ray tracing device
+      gprtBuildShaderBindingTable(context, GPRT_SBT_RAYGEN);
+    }
+
+    // update time to move primitives. then, rebuild accel.
+    gprtComputeSet1f(vertexProgram, "now", float(gprtGetTime(context)));
+    gprtBuildShaderBindingTable(context, GPRT_SBT_COMPUTE);
+    gprtComputeLaunch1D(context, vertexProgram, numTriangles);
+    gprtAccelBuild(context, trianglesAccel);
+    gprtAccelBuild(context, world);
```

Comment:
These need to happen in a certain order? Or would the `world` call handle it if the `trianglesAccel` all hasn't occurred yet?

Reply:
these need to happen in a certain order yep. The bottom level tree must be built before the top level tree can be built, since the top level tree stores device addresses at the leaves to the built bottom level trees. 

It's possible that we could make gprtAccelBuild build all un-built bottom level trees, but checking whether or not each bottom level tree is built is expensive, and closes the doors to a couple optimizations I have in mind down the road with GPU-parallel instance generation.

Reply:
For what it's worth, the OWL API we're based off of also makes a similar assumption that all bottom level trees be built before building the top level tree.

Reply:
Just to give an idea, we might write a GPRT compute kernel that fills an instance "primitive" buffer with a different set of instances depending on the level-of-detail we want for an object. Far away objects from the camera might use low level-of-detail instances, while closer objects would receive higher level of detail instances. This LOD selection would happen in parallel on the device to make it fast. 

If we require that the top level tree builds all un-built bottom level trees, we would need to download all BLAS references in this primitive buffer from the device back to the host and then iterate over these BLAS in series, checking to see if each was built. The download of those references is the pain point i'm trying to avoid by instead throwing it on the user to make sure each BLAS is pre-built before going into the TLAS.

Reply:
Makes sense. I think a comment here as a reminder would be good if you're up for it. 

Maybe a comment on order of object destruction if needed too

Reply:
thinking i'll just do the comment for building, since all the previous examples have established destruction order already.

Reply:
Comments for order of destruction? What do you think?